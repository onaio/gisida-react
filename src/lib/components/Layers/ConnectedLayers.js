/**
 * Handle display of submenu and layer items by utilizing redux state.
 * Keeps track of open submenus in redux state
 * 
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Layer from '../Layer/Layer';
import { connect } from 'react-redux';
import { Actions } from 'gisida';
const mapStateToProps = (state, ownProps) => {
  const MAP = state[ownProps.mapId];
  const { mapId, layers, currentRegion, preparedLayers } = ownProps;
  return {
    openGroups: MAP.openGroups,
    mapId,
    layers,
    currentRegion,
    preparedLayers,
  };
};
export class ConnectedLayers extends Component {
  /**
   * Toggle open and close a menu group
   * @param {Object} e Event
   * @param {Object} layer Group to be toggled
   * @param {number} groupCount Group count
   */
  toggleSubMenu(e, layer, groupCount) {
    e.preventDefault();
    const { openGroups } = this.props;
    let index = -1;
    let count = 0;
    let found = false;
    while (!found && count < openGroups.length) {
      if (openGroups[count].count === groupCount && openGroups[count].group === layer) {
        index = count;
        found = true;
      }
      count += 1;
    }
    this.props.dispatch(Actions.toggleGroups(this.props.mapId, layer, index, false, groupCount));
  }
  /**
   * Get the group open state
   * @param {number} groupCount Group count
   * @param {string} groupName Group name
   * @returns {boolean} Whether the group is open or not
   */
  isGroupOpen(groupCount, groupName) {
    return (
      this.props.openGroups &&
      this.props.openGroups.filter(group => group.count === groupCount && group.group === groupName)
        .length
    );
  }
  render() {
    const { mapId, layers, currentRegion, preparedLayers, auth } = this.props;
    let layerKeys;
    let layerObj;
    let layerItem = [];
    const subLayerIds = [];
    if (!preparedLayers) {
      return false;
    }
    layerKeys = Object.keys(preparedLayers);
    for (let lo = 0; lo < layerKeys.length; lo += 1) {
      layerObj = this.props.preparedLayers[layerKeys[lo]];
      if (layerObj.layers) {
        for (let s = 0; s < layerObj.layers.length; s += 1) {
          subLayerIds.push(layerObj.layers[s]);
        }
      }
    }
    layers.forEach(layer => {
      if (
        (!currentRegion || (layer.region && layer.region === currentRegion)) &&
        !subLayerIds.includes(layer.id)
      ) {
        if (layer.id && (!auth || !auth.authConfigs)) {
          if (this.props.layerItem) {
            const CustomLayerItem = this.props.layerItem;
            layerItem.push(<CustomLayerItem key={layer.id} mapId={mapId} layer={layer} />);
          } else {
            layerItem.push(<Layer key={layer.id} mapId={mapId} layer={layer} />);
          }
        } else if (layer.id && auth) {
          const { authConfigs, userInfo } = auth;
          let activeId = layer.id;
          let users;
          if (activeId.indexOf('.json') !== -1) {
            // layer ids from github shared repo have .json suffix
            // split to remove .json part
            activeId = activeId.split('.')[0];
          }
          const LocalAuthConfig = JSON.parse(localStorage.getItem('authConfig'));
          users = authConfigs.LAYERS && authConfigs.LAYERS[activeId]; // list of users with access to the layer
          authConfigs.LAYERS = authConfigs.LAYERS || LocalAuthConfig.LAYERS;
          users = authConfigs.LAYERS[activeId]; 
          // list of users with access to the layer
          // check if logged in user exists in the list of users
          // who have access to the layer
          if (
            (users && userInfo && users.includes(userInfo.username)) ||
            (authConfigs.LAYERS &&
              authConfigs.LAYERS.ALL &&
              authConfigs.LAYERS.ALL.includes(userInfo.username))
          ) {
            if (this.props.layerItem) {
              const CustomLayerItem = this.props.layerItem;
              layerItem.push(<CustomLayerItem key={layer.id} mapId={mapId} layer={layer} />);
            } else {
              layerItem.push(<Layer key={layer.id} mapId={mapId} layer={layer} />);
            }
          }
        } else {
          Object.keys(layer).forEach((d, i) => {
            layerItem = layerItem.concat([
              <li>
                <a
                  key={`${d}-${i}-link`}
                  className="sub-category"
                  onClick={e => this.toggleSubMenu(e, d, layer[d].count)}
                >
                  {d}
                  <span
                    className={`category glyphicon glyphicon-chevron-${
                      this.isGroupOpen(layer[d].count, d) ? 'down' : 'right'
                    }`}
                  />
                </a>
              </li>,
              this.isGroupOpen(layer[d].count, d) ? (
                <ConnectedLayers
                  openGroups={this.props.openGroups}
                  dispatch={this.props.dispatch}
                  layerItem={this.props.layerItem}
                  key={`${d}-${i}`}
                  mapId={mapId}
                  layers={layer[d].layers}
                  currentRegion={currentRegion}
                  preparedLayers={preparedLayers}
                  auth={this.props.auth}
                />
              ) : null,
            ]);
          });
        }
      }
      return null;
    });
    return <ul className="layers">{layerItem}</ul>;
  }
}
ConnectedLayers.propTypes = {
  mapId: PropTypes.string.isRequired,
  layers: PropTypes.arrayOf(PropTypes.any).isRequired,
  preparedLayers: PropTypes.arrayOf(PropTypes.any),
  currentRegion: PropTypes.string,
  auth: PropTypes.object
};
export default connect(mapStateToProps)(ConnectedLayers);