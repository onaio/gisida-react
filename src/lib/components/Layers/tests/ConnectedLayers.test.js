import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ConnectedLayers from './../ConnectedLayers'
import layerObj from './fixtures/sample-layer.json';
import configureMockStore from "redux-mock-store";
import { Provider } from 'react-redux';
/** Mock store */

const mockStore = configureMockStore();
const initStore = {
  MAP: {
      openGroups: [
        {
          "count": 0,
          "group": "Demographics"
        },
        {
          "count": 1,
          "group": "Crop production"
        }
      ]
  }
}
const store = mockStore(initStore);
describe('Layers', () => {
  layerObj.id = 'sample-layer';
  const layers = [layerObj];
  const preparedLayers = {'sample-layer-id': layerObj}
  
  const componentWrapper = shallow(
    <Provider store={store}>
        <ConnectedLayers
        mapId={'map-1'}
        layers={layers}
        currentRegion={undefined}
        preparedLayers={preparedLayers}
        />
    </Provider>
  );
  it('component renders correctly', () => {
    const json = toJson(componentWrapper)
    expect(json).toMatchSnapshot();
  });
});
