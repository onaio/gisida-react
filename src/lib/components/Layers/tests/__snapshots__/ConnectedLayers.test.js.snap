// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layers component renders correctly 1`] = `
<Connect(ConnectedLayers)
  layers={
    Array [
      Object {
        "category": "Boundaries and Place Labels",
        "id": "sample-layer",
        "label": "Boundaries",
        "paint": Object {
          "line-color": "#444",
          "line-opacity": 0.6,
          "line-width": 1,
        },
        "source": Object {
          "layer": "som_adm_2-0j5bly",
          "type": "vector",
          "url": "mapbox://ona.6gl2shzi",
        },
        "type": "line",
        "visible": true,
      },
    ]
  }
  mapId="map-1"
  preparedLayers={
    Object {
      "sample-layer-id": Object {
        "category": "Boundaries and Place Labels",
        "id": "sample-layer",
        "label": "Boundaries",
        "paint": Object {
          "line-color": "#444",
          "line-opacity": 0.6,
          "line-width": 1,
        },
        "source": Object {
          "layer": "som_adm_2-0j5bly",
          "type": "vector",
          "url": "mapbox://ona.6gl2shzi",
        },
        "type": "line",
        "visible": true,
      },
    }
  }
/>
`;
